R/run-nmtran.R

```diff
@@ -0,0 +1,330 @@
+          "i" = "Please ensure only one version is set to the default"
+        )
+      )
+    }else if(length(default_nm) == 1){
+      default_nm <- default_nm[[1]]
+    }else{
+      # If no default, use the last one (likely higher version)
+      default_nm <- nm_config[[length(nm_config)]]
+    }
+  }
+
+  # Set NM-TRAN executable path
+  nm_path <- default_nm$home
+  nmtran_exe <- file.path(nm_path, "tr", "NMTRAN.exe")
+
+  # Check if nmtran_presort exists
+  presort_dir <- file.path(dirname(dirname(nmtran_exe)), "util")
+  nmtran_presort_exe <- file.path(presort_dir, "nmtran_presort")
+  run_presort <- unname(fs::file_exists(nmtran_presort_exe))
+
+  return(
+    list(
+      nmtran_exe = nmtran_exe,
+      nonmem_version = basename(default_nm$home),
+      cmd_args = cmd_args,
+      nmtran_presort_exe = if(run_presort) nmtran_presort_exe else NULL
+    )
+  )
+}
+
+#' Execute `NM-TRAN` in a given directory
+#'
+#' Execute `NM-TRAN` in a given directory. Also runs `nmtran_presort` if an
+#' executable is found.
+#'
+#' @param nmtran_exe Path to `NM-TRAN` executable.
+#' @param mod_path Path of a model to evaluate. Should be relative to `dir`.
+#' @param cmd_args A character vector of command line arguments for the `NM-TRAN`
+#'  execution call
+#' @param nmtran_presort_exe Path to `nmtran_presort` executable. Only available
+#' for `NONMEM` versions `nm74gf`, `nm74gf_nmfe`, and `nm75`. If provided, will run
+#' `nmtran_presort` before running `NM-TRAN`. Set to `NULL` to skip this step.
+#' @param dir Directory in which to execute the command.
+#'
+#' @keywords internal
+execute_nmtran <- function(
+    nmtran_exe,
+    mod_path,
+    cmd_args = c("0", "0", "2"),
+    nmtran_presort_exe = NULL,
+    dir = "."
+){
+  checkmate::assert_directory_exists(dir)
+  checkmate::assert_character(cmd_args, len = 3)
+
+  run_dir <- as.character(fs::path_real(dir))
+
+  # Check if nmtran_presort exists
+  run_presort <- !is.null(nmtran_presort_exe)
+
+  # Preprocess with nmtran_presort
+  if(run_presort){
+    presort.p <- processx::run(
+      command = nmtran_presort_exe, wd = run_dir, args = character(),
+      stdout = "|", stdin = file.path(run_dir, mod_path),
+      stderr_to_stdout = TRUE, error_on_status = FALSE
+    )
+
+    presort_status_val <- presort.p$status
+    if(is.na(presort_status_val)){
+      rlang::abort("nmtran_presort terminated unexpectedly")
+    }else if(presort_status_val != 0){
+      return(
+        list(
+          nmtran_model = file.path(run_dir, mod_path),
+          run_dir = run_dir,
+          status = "nmtran_presort failed. See errors.",
+          status_val = presort_status_val,
+          output_lines = presort.p$stdout
+        )
+      )
+    }
+
+    # Write the output to tempzzzz1 control stream
+    writeLines(presort.p$stdout, con = file.path(run_dir, "tempzzzz1.ctl"))
+  }
+
+  # Run NM-TRAN
+  stdin_file <- ifelse(run_presort, "tempzzzz1.ctl", mod_path)
```

Comment:
This is operating where the input and output should always be a scalar, so please don't use the vectorized `ifelse`.